common/core/desktop/src/kmx/kmx_processor.cpp

```diff
@@ -76,19 +80,33 @@ KMX_BOOL KMX_Processor::ProcessEvent(km_kbp_state *state, KMX_UINT vkey, KMX_DWO
 
   m_kbp_state = state;
 
-  if (m_environment.capsLock())
-    modifiers |= CAPITALFLAG;
+  m_environment.Load(KM_KBP_KMX_ENV_CAPSLOCK, (modifiers & CAPITALFLAG) ? u"1" : u"0");
+
+  ResetCapsLock(modifiers);
 
   m_state.vkey = vkey;
   m_state.charCode = VKeyToChar(modifiers, vkey);
   m_modifiers = modifiers;
   m_state.LoopTimes = 0;
 
-  if (kbd->StartGroup[BEGIN_UNICODE] == (KMX_DWORD) -1) {
+  KMX_BOOL isKeyUp = (modifiers & KM_KBP_MODIFIER_MASK_RELEASE) > 0;
+
+  if (kbd->StartGroup[BEGIN_UNICODE] == (KMX_DWORD)-1)
+  {
     DebugLog("Non-Unicode keyboards are not supported.");
     return FALSE;
   }
 
+  switch (vkey)
+  {
+    case KM_KBP_VKEY_CAPS:
+      KeyCapsLockPress(isKeyUp);
+      break;
+    case KM_KBP_VKEY_SHIFT:
+      KeyShiftPress(isKeyUp);
+      break;
+    }
+
   LPGROUP gp = &kbd->dpGroupArray[kbd->StartGroup[BEGIN_UNICODE]];
```

Comment:
If we pass through `isKeyUp`, that needs to be tested so we don't run through the normal rule processing -- otherwise we're gonna get hoopy things on the key up events, i.e. around line 110 we add:

```cpp
if(isKeyUp) {
  m_kbp_state = nullptr;
  return FALSE;
}
```

Reply:
done